etna/packages/etna-py/src/mountetna/magma.py

```diff
@@ -0,0 +1,320 @@
+                            record_name,
+                            attr.link_model_name,
+                            self.update_record(attr.link_model_name, table_id),
+                            attr_name
+                        )
+
+    def __len__(self):
+        return sum(len(rev) for rev in self.revisions.values())
+
+    def empty(self):
+        return len(self) == 0
+
+    def extend(self, other: "UpdateRequest"):
+        for model_name, docs in other.revisions.items():
+            for record_name, revision in docs.items():
+                self.update_record(model_name, record_name, revision)
+
+    # TODO: Add more basic validations?
+    def validate(self, models: Dict[str, Model]) -> typing.Iterable[str]:
+        for model_name, docs in self.revisions.items():
+            if model_name not in models:
+                yield f"Model '{model_name}' does not exist"
+                continue
+
+            template = models[model_name].template
+            for record_name, revision in docs.items():
+                for attr, value in revision.items():
+                    if attr not in template.attributes:
+                        yield f"Model '{model_name}', Attribute '{attr}' does not exist"
+                        continue
+
+    def update_record(
+        self, model_name: str, record_name: str, attrs: Dict[str, typing.Any] = dict()
+    ) -> Dict[str, typing.Any]:
+        attrs = {k: normalize_for_magma(v) for k, v in attrs.items()}
+        record = self.revisions.setdefault(model_name, {}).setdefault(record_name, {})
+        record.update(**attrs)
+        return record
+
+    def append_table(
+        self,
+        parent_model_name: str,
+        parent_record_name: str,
+        model_name: str,
+        attrs: Dict[str, typing.Any],
+        attribute_name: str,
+    ):
+        parent_revision = self.update_record(parent_model_name, parent_record_name, {})
+        table = parent_revision.setdefault(attribute_name, [])
+        id = f"::{model_name}{len(self.revisions.setdefault(model_name, {})) + 1}"
+        table.append(id)
+        self.update_record(model_name, id, attrs)
+        return id
+
+@serialize
+@deserialize
+@dataclasses.dataclass
+class RetrievalResponse:
+    models: Dict[str, Model] = dataclasses.field(default_factory=dict)
+
+    def empty(self):
+        for model in self.models.values():
+            if not model.empty():
+                return False
+        return True
+
+    def extend(self, other: "RetrievalResponse"):
+        for model_name, model in other.models.items():
+            self.models.setdefault(model_name, Model()).extend(model)
+
+@serialize
+@deserialize
+@dataclasses.dataclass
+class QueryRequest:
+    query: List[typing.Any] = dataclasses.field(default_factory=list)
+    project_name: str = ""
+
+@serialize
+@deserialize
+@dataclasses.dataclass
+class QueryResponse:
+    answer: List[typing.Any] = dataclasses.field(default_factory=list)
```

Comment:
I'm not sure how to correctly type-hint a nested list of lists or strings...

Reply:
perhaps `List[List[str]]`

Reply:
ah, yes.  Perhaps `List[Union[str, List[str]]` then?  (`from typing import Union`)

Reply:
Hmm, apparently something like: `Query = List[Union[ str, 'Query' ] ]` then `query : Query` using https://peps.python.org/pep-0484/#forward-references